CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1599 @@
+cannot choose to supply alternative implementations of them.
+
+### In-service upgrade
+
+Long-lived contracts may need upgrades and bug fixes during their
+lifetimes. This is especially true of contracts made up of many
+modules--every time a dependency is upgraded or receives a bug fix,
+the question of whether or not to upgrade the client contract
+arises. However, the problem of upgrading contracts *securely* is a
+tricky one, and exists whether or not modules are used. Therefore this
+CIP does not address this problem specifically: developers should use
+the same mechanisms to handle upgrades of scripts with dependencies,
+as they use to upgrade scripts without dependencies. The only thing we
+note here is that the need for upgrades is more likely to arise when a
+script depends on many others, so it is more important to be prepared
+for it. Note that, because a script contains the hashes of its
+dependencies, no 'silent' upgrade can occur: the hash of a script
+depends on *all* of its code, including the code of its dependencies,
+so any change in a dependency will lead to a change in the script
+hash.
+
+### Lazy loading
+
+The 'lazy loading' variation in the specification section above
+permits fewer modules to be supplied in a transaction.  Dependency
+trees have a tendency to grow very large; when one function in a
+module uses another module, it becomes a dependency of the entire
+module and not just of that function. It is easy to imagine situations
+in which a script depends on many modules, but a particular call
+requires only a few of them. For example, if a script offers a choice
+of protocols for redemption, only one of which is used in a particular
+call, then many modules may not actually be needed. The variation
+allows a transaction to omit the unused modules in such cases. This
+reduces the size of the transaction, which need provide fewer
+witnesses, but more importantly it reduces the amount of code which
+must be loaded from reference UTxOs.
+
+If a script execution *does* try to use a module which was not
+provided, it will encounter a run-time type error and fail (unless the
+module value was `builtin unit`, in which case the script will behave
+as though the module had been provided).
+
+To take advantage of this variation, it is necessary, when a
+transaction is constructed, to *observe* which script arguments are
+actually used by the script invocations needed to validate the
+transaction. The transaction balancer runs the scripts anyway, and so
+can in principle observe the uses of script arguments, and include
+witnesses in the transaction for just those arguments that are used.
+
+#### Balancer modifications
+
+To take advantage of 'lazy loading', it's necessary to identify
+reference scripts that are *dynamically* unused, when the scripts in a
+transaction run. The best place to do that is in a transaction
+balancer, which needs to run the scripts anyway, both to check that

```

Comment:
I haven't come across the term "transaction balancer" before, and Google isn't being very helpful.  To be clear,  this is  something that is run **off the chain** prior to transaction submission, no?  If that's the case, I'm not sure if it's necessary to go into this level of detail in the CIP.  Since on-chain scripts are deterministic it seems pretty clear that you can run them in some kind of instrumented evaluator to determine which parts of the AST are actually required.

Reply:
It is basically the code for building transactions, e.g., in cardano-api there is `Cardano.Api.Fees.makeTransactionBodyAutoBalance`. Yes it runs off-chain.

Reply:
Yes, indeed, it's an off-chain step that is part of preparing a transaction for submission. Should transaction balancing be addressed in the CIP? I'd agree that it's not *necessary* to do so, in that it doesn't affect the chain itself, but I've included this discussion in an effort to be helpful--after all, *somebody* is going to have to implement script dropping if lazy loading is to be valuable, and--as I think the discussion shows--it's not totally straightforward to do that. So no harm in describing some options.